aws-xray-recorder-sdk-core/src/test/java/com/amazonaws/xray/ConcurrencyTest.java

```diff
@@ -64,4 +70,31 @@ void testManyThreads() throws InterruptedException {
         }
     }
 
+    @Test
+    public void testPrecursorIdConcurrency() throws Exception {
+        Segment seg = new SegmentImpl(AWSXRay.getGlobalRecorder(), "test");
+        SubsegmentImpl subseg = new SubsegmentImpl(AWSXRay.getGlobalRecorder(), "test", seg);
+
+        final long startTime = System.currentTimeMillis();
+
+        Thread thread1 = new Thread(() -> {
+            while (true) {
+                subseg.addPrecursorId("ID" + new Random().nextInt());
+            }
+        });
+        thread1.start();
+
+        Callable<Void> callable = (Callable) () -> {
+            while (System.currentTimeMillis() - startTime < TimeUnit.SECONDS.toMillis(1)) {
+                Iterator it = subseg.getPrecursorIds().iterator();
```

Comment:
I think this `Callable` should also be using `addPrecursorId` so this function is tested being called in 2 threads.

Reply:
I was able to produce a concurrent modification exception with this code.
Iterating and changing the data structure is not allowed in a normal HashSet

If they are both adding, then it has to be the same bucket in the hash table at the same time which is harder to get to occur.